crates/router/src/connector/utils.rs

```diff
@@ -47,26 +90,74 @@ impl CardData for api::CCard {
         self.card_cvc.peek().clone()
     }
 }
-impl PaymentsRequestData for types::PaymentsAuthorizeRouterData {
-    fn get_attempt_id(&self) -> Result<String, Error> {
-        self.attempt_id
+pub trait PhoneDetailsData {
+    fn get_number(&self) -> Result<Secret<String>, Error>;
+    fn get_country_code(&self) -> Result<String, Error>;
+}
+
+impl PhoneDetailsData for api::PhoneDetails {
+    fn get_country_code(&self) -> Result<String, Error> {
+        self.country_code
             .clone()
-            .ok_or_else(missing_field_err("attempt_id"))
+            .ok_or_else(missing_field_err("billing.phone.country_code"))
     }
-
-    fn get_billing_country(&self) -> Result<String, Error> {
-        self.address
-            .billing
+    fn get_number(&self) -> Result<Secret<String>, Error> {
+        self.number
             .clone()
-            .and_then(|a| a.address)
-            .and_then(|ad| ad.country)
-            .ok_or_else(missing_field_err("billing.country"))
+            .ok_or_else(missing_field_err("billing.phone.number"))
     }
+}
 
-    fn get_card(&self) -> Result<api::CCard, Error> {
-        match self.request.payment_method_data.clone() {
-            api::PaymentMethod::Card(card) => Ok(card),
-            _ => Err(missing_field_err("card")()),
-        }
+pub trait AddressDetailsData {
+    fn get_first_name(&self) -> Result<&Secret<String>, Error>;
+    fn get_last_name(&self) -> Result<&Secret<String>, Error>;
+    fn get_line1(&self) -> Result<&Secret<String>, Error>;
+    fn get_city(&self) -> Result<&String, Error>;
+    fn get_line2(&self) -> Result<&Secret<String>, Error>;
+    fn get_zip(&self) -> Result<&Secret<String>, Error>;
+    fn get_country(&self) -> Result<&String, Error>;
+}
+
+impl AddressDetailsData for api::AddressDetails {
+    fn get_first_name(&self) -> Result<&Secret<String>, Error> {
```

Comment:
yes in such cases we don't even need to use this util because this util is used only when the field is mandatory.

Example 1: when firstname is optional
`address.first_name`
Example 2: when firstname is mandatory
`address.get_first_name()`